src/machines/java/com/enderio/machines/common/blockentity/task/CraftingMachineTask.java

```diff
@@ -0,0 +1,228 @@
+        return recipe;
+    }
+
+    // region Abstract Implementation
+
+    protected abstract void consumeInputs(R recipe);
+
+    protected abstract int makeProgress(int remainingProgress);
+
+    protected abstract int getProgressRequired(R recipe);
+
+    // endregion
+
+    // region Task Implementation
+
+    @Override
+    public void tick() {
+        // If the recipe is done, don't let it tick.
+        if (isComplete)
+            return;
+
+        // If the recipe failed to load somehow, cancel
+        if (recipe == null) {
+            isComplete = true;
+            return;
+        }
+
+        // Get the outputs list.
+        if (!hasDeterminedOutputs) {
+            hasDeterminedOutputs = true;
+            outputs = recipe.craft(container, level.registryAccess());
+
+            // TODO: Compact any items that are the same into singular stacks?
+        }
+
+        // If we can't input or output, cancel the task. However, if for some reason we can't output after the inputs are collected, don't.
+        if (!hasConsumedInputs && (!placeOutputs(outputs, true) || !recipe.matches(container, level))) {
+            isComplete = true;
+            // This means if a sag mill recipe outputs 2 it cancels the recipe, and the determined outputs are cleared. It's a weird behaviour but not necessarily a bug.
+            // We might want to review how this works in the future, as right now we wait for an inventory change rather than the machine tick repeatedly.
+            return;
+        }
+
+        // If we haven't done so already, consume inputs for the recipe.
+        if (!hasConsumedInputs) {
+            // Consume inputs for the recipe.
+            consumeInputs(recipe);
+            hasConsumedInputs = true;
+
+            // Store the recipe energy cost.
+            // This is run afterward as it allows container context changes after takeInputs()
+            progressRequired = getProgressRequired(recipe);
+        }
+
+        // Try to consume as much energy as possible to finish the craft.
+        if (progressMade < progressRequired) {
+            progressMade += makeProgress(progressRequired - progressMade);
+        }
+
+        // If the recipe has been crafted, attempt to put it into storage
+        if (progressMade >= progressRequired) {
+            // Attempt to complete the craft
+            if (placeOutputs(outputs, false)) {
+                // The receiver was able to take the outputs, task complete.
+                isComplete = true;
+            }
+        }
+    }
+
+    @Override
+    public float getProgress() {
+        if (recipe == null)
+            return 0.0f;
+        return progressMade / (float) progressRequired;
+    }
+
+    @Override
+    public boolean isComplete() {
+        return isComplete;
+    }
+
+    // endregion
+
+    // region Resource Depletion
+
+    protected boolean placeOutputs(List<OutputStack> outputs, boolean simulate) {
+        // TODO: Handle fluids too.
```

Comment:
At some point I might make a "MachineStorages" class/interface or something that will allow fetching of inventories, tanks or any other kinds of resource targets, rather than having to pass the inventory, tank etc. every time individually.